plexus-compilers/plexus-compiler-aspectj/src/main/java/org/codehaus/plexus/compiler/ajc/AspectJCompiler.java

```diff
@@ -565,75 +572,164 @@ private void checkForAspectJRT( List<String> cp )
     private void setSourceVersion( AjBuildConfig buildConfig, String sourceVersion )
         throws CompilerException
     {
-        if ( "11".equals( sourceVersion ) )
-        {
-            buildConfig.getOptions().sourceLevel = ClassFileConstants.JDK11;
-        }
-        else if ( "10".equals( sourceVersion ) )
-        {
-            buildConfig.getOptions().sourceLevel = ClassFileConstants.JDK10;
-        }
-        else if ( "9".equals( sourceVersion ) )
-        {
-            buildConfig.getOptions().sourceLevel = ClassFileConstants.JDK9;
-        }
-        else if ( "1.9".equals( sourceVersion ) )
-        {
-            buildConfig.getOptions().sourceLevel = ClassFileConstants.JDK9;
-        }
-        else if ( "1.8".equals( sourceVersion ) )
-        {
-            buildConfig.getOptions().sourceLevel = ClassFileConstants.JDK1_8;
-        }
-        else if ( "1.7".equals( sourceVersion ) )
+        buildConfig.getOptions().sourceLevel = versionStringToMajorMinor( sourceVersion );
+    }
+
+    /**
+     * Set the target version in ajc compiler
+     *
+     * @param buildConfig
+     * @param targetVersion
+     */
+    private void setTargetVersion( AjBuildConfig buildConfig, String targetVersion )
+        throws CompilerException
+    {
+        buildConfig.getOptions().targetJDK = versionStringToMajorMinor( targetVersion );
+    }
+
+    private static final Map<String, Long> VERSION_MAP = new HashMap<>();
+
+    static
+    {
+        // Default in Maven Compiler 3.8.1
+        VERSION_MAP.put( "",     JDK1_6 );
+
+        VERSION_MAP.put( "1.1",  JDK1_1 );
+        VERSION_MAP.put( "1.2",  JDK1_2 );
+        VERSION_MAP.put( "1.3",  JDK1_3 );
+        VERSION_MAP.put( "1.4",  JDK1_4 );
+        VERSION_MAP.put( "1.5",  JDK1_5 );
+        VERSION_MAP.put( "5",    JDK1_5 );
+        VERSION_MAP.put( "5.0",  JDK1_5 );
+        VERSION_MAP.put( "1.6",  JDK1_6 );
+        VERSION_MAP.put( "6",    JDK1_6 );
+        VERSION_MAP.put( "6.0",  JDK1_6 );
+        VERSION_MAP.put( "1.7",  JDK1_7 );
+        VERSION_MAP.put( "7",    JDK1_7 );
+        VERSION_MAP.put( "7.0",  JDK1_7 );
+        VERSION_MAP.put( "1.8",  JDK1_8 );
+        VERSION_MAP.put( "8",    JDK1_8 );
+        VERSION_MAP.put( "8.0",  JDK1_8 );
+        VERSION_MAP.put( "1.9",  JDK9 );
+        VERSION_MAP.put( "9",    JDK9 );
+        VERSION_MAP.put( "9.0",  JDK9 );
+        VERSION_MAP.put( "10",   JDK10 );
+        VERSION_MAP.put( "10.0", JDK10 );
+        VERSION_MAP.put( "11",   JDK11 );
+        VERSION_MAP.put( "11.0", JDK11 );
+        VERSION_MAP.put( "12",   JDK12 );
+        VERSION_MAP.put( "12.0", JDK12 );
+        VERSION_MAP.put( "13",   JDK13 );
+        VERSION_MAP.put( "13.0", JDK13 );
+        VERSION_MAP.put( "14",   JDK14 );
+        VERSION_MAP.put( "14.0", JDK14 );
+        VERSION_MAP.put( "15",   JDK15 );
+        VERSION_MAP.put( "15.0", JDK15 );
+        VERSION_MAP.put( "16",   JDK16 );
+        VERSION_MAP.put( "16.0", JDK16 );
```

Comment:
Haven't checked, but I think there is a JavaVersion class somewhere in the class path.
Also, do we need the `.0` versions starting from 9? I never encountered them.
And while at it, you can remove 5, 5.0, 6, 6.0, 7, 7.0.

Reply:
> Also, do we need the .0 versions starting from 9? I never encountered them.

AspectJ is an ECJ fork, and [ECJ supports those version numbers](https://github.com/eclipse/eclipse.jdt.core/blob/6c5346e6b20e0e2cfc2fc914b5d2eaa9194e3914/org.eclipse.jdt.core/batch/org/eclipse/jdt/internal/compiler/batch/messages.properties#L239-L258), hence AspectJ does too.

Actually, this can go away in the future, if I switch from the internal AJDT compiler interface to AJC being a subclass of ECJ, merely extending instead of redefining it.

> And while at it, you can remove 5, 5.0, 6, 6.0, 7, 7.0.

Those are legal for ECJ/AJC too, see above.

BTW, that verbose way to create a map is ugly - still better than the if-else (or was it swirtch-case) from before, but if on the class path there is a tool class somewhere, permitting me to do this in a more compact way than using lots of `put` statements, please let me know. I just wanted to get it working and be reusable for all of source, target and release parameters in a uniform way. Before, only source was covered.

Reply:
org.codehaus.plexus.languages.java.version.JavaVersion is the solution to reduce this list, which will turn it back into if/switch statements :)

Reply:
`plexus-languages` is not on the classpath. Is it worth pulling it in for this?

Reply:
OK, I can answer my own question already: It does not make sense. First of all, it cannot parse `8.0` correctly into a major `8` (still returns major `8.0`), but ECJ/AJC support it. Also comparing versions does not work correctly (e.g. `JavaVersion.parse("8").isBefore("1.9")` returns `false)`. Furthermore, there would zero benefit, even if that utility worked perfectly, because I need to convert strings into AJDT `ClassFileConstants.*` values of type `long`, encoding major-minor in a way not provided by the Plexus `JavaVersion`, e.g. `ClassFileConstants.JDK1_8` evaluates to `3407872`.

so I think we can scratch this and leave it as it is for now. Like I already explained, if in the future any developer decides to switch from the ADJT interfact to the AJC batch compiler, we can let the compiler do the parsing. But the AJDT API is more strongly typed, not dependent on only parsing strings from a CLI. So for now this is how it is.

Reply:
This is how it could be used:

       // only need for first digit 
       assertThat( JavaVersion.parse( "8.0" ).getValue( 1 ), is( "8" ) );
       // compare with the major version to support 1.x values
       assertThat( JavaVersion.parse( "1.8" ).asMajor().isBefore( JavaVersion.parse( "1.9" ).asMajor() ), is( true ) );       
       assertThat( JavaVersion.parse( "1.8" ).asMajor().isBefore( JavaVersion.parse( "9" ).asMajor() ), is( true ) );
       assertThat( JavaVersion.parse( "8" ).asMajor().isBefore( JavaVersion.parse( "1.9" ).asMajor() ), is( true ) );       
       assertThat( JavaVersion.parse( "8" ).asMajor().isBefore( JavaVersion.parse( "9" ).asMajor() ), is( true ) );

Reply:
Yeah, nice, but how does that help me here?